examples/src/bin/unbuffered-client.rs

```diff
@@ -0,0 +1,307 @@
+        .with_protocol_versions(&[&TLS13])
+        .unwrap()
+        .with_root_certificates(build_root_store()?)
+        .with_no_client_auth();
+    config.enable_early_data = SEND_EARLY_DATA;
+
+    let config = Arc::new(config);
+
+    let mut incoming_tls = [0; INCOMING_TLS_BUFSIZ];
+    let mut outgoing_tls = vec![0; OUTGOING_TLS_INITIAL_BUFSIZ];
+
+    converse(&config, false, &mut incoming_tls, &mut outgoing_tls)?;
+    if SEND_EARLY_DATA {
+        eprintln!("---- second connection ----");
+        converse(&config, true, &mut incoming_tls, &mut outgoing_tls)?;
+    }
+
+    Ok(())
+}
+
+fn converse(
+    config: &Arc<ClientConfig>,
+    send_early_data: bool,
+    incoming_tls: &mut [u8],
+    outgoing_tls: &mut Vec<u8>,
+) -> Result<(), Box<dyn Error>> {
+    let mut conn = UnbufferedClientConnection::new(Arc::clone(config), SERVER_NAME.try_into()?)?;
+    let mut sock = TcpStream::connect(format!("{SERVER_NAME}:{PORT}"))?;
+
+    let mut incoming_used = 0;
+    let mut outgoing_used = 0;
+
+    let mut open_connection = true;
+    let mut sent_request = false;
+    let mut received_response = false;
+    let sent_early_data = false;
+
+    let mut iter_count = 0;
+    while open_connection {
+        let UnbufferedStatus { mut discard, state } =
+            conn.process_tls_records(&mut incoming_tls[..incoming_used])?;
+
+        match dbg!(state) {
+            ConnectionState::AppDataAvailable(mut state) => {
+                while let Some(res) = state.next_record() {
+                    let AppDataRecord {
+                        discard: new_discard,
+                        payload,
+                    } = res?;
+                    discard += new_discard;
+
+                    if payload.starts_with(b"HTTP") {
+                        let response = core::str::from_utf8(payload)?;
+                        let header = response
+                            .lines()
+                            .next()
+                            .unwrap_or(response);
+
+                        println!("{header}");
+                    } else {
+                        println!("(.. continued HTTP response ..)");
+                    }
+
+                    received_response = true;
+                }
+            }
+
+            ConnectionState::MustEncodeTlsData(mut state) => {
+                try_or_resize_and_retry(
+                    |out_buffer| state.encode(out_buffer),
+                    |e| {
+                        if let EncodeError::InsufficientSize(is) = &e {
+                            Ok(*is)
+                        } else {
+                            Err(e.into())
+                        }
+                    },
+                    outgoing_tls,
+                    &mut outgoing_used,
+                )?;
+            }
+
+            ConnectionState::MustTransmitTlsData(mut state) => {
+                // if let Some(mut may_encrypt_early_data) = state.may_encrypt_early_data() {
```

Comment:
this gets uncommented in a later commit and can't be commented out in this commit because it uses an API that does not yet exist. I'll remove the commented code when tidying up the commit history.